programmers/level1/[1차]-다트-게임/김은경.js

```diff
@@ -0,0 +1,78 @@
+function solution(dartResult) {
+  const pattern = /[0-9]+[STD][*#]?/g;
+  const array = dartResult.match(pattern);
+  const score = [];
+
+  array.forEach((el) => {
+    const individual = el.split('');
+    let divider = individual[individual.length - 1];
+    if (divider === 'S' || divider === 'T' || divider === 'D') {
+      const gottonNum = parseInt(individual.slice(0, -1).join(''));
+      let getScore = 0;
+      switch (divider) {
+        case 'S':
+          getScore = gottonNum;
+          break;
+        case 'D':
+          getScore = Math.pow(gottonNum, 2);
+          break;
+        case 'T':
+          getScore = Math.pow(gottonNum, 3);
+          break;
+      }
+      score.push(getScore);
+    }
+    if (divider === '*') {
+      score[score.length - 1] *= 2;
+      let divider = individual[individual.length - 2];
+      const gottonNum = parseInt(individual.slice(0, -2).join(''));
+      let getScore = 0;
+      switch (divider) {
+        case 'S':
+          getScore = 2 * gottonNum;
+          break;
+        case 'D':
+          getScore = 2 * Math.pow(gottonNum, 2);
+          break;
+        case 'T':
+          getScore = 2 * Math.pow(gottonNum, 3);
+          break;
+      }
+      score.push(getScore);
+    }
+    if (divider === '#') {
+      let divider = individual[individual.length - 2];
+      const gottonNum = parseInt(individual.slice(0, -2).join(''));
+      let getScore = 0;
+      switch (divider) {
+        case 'S':
+          getScore = -1 * gottonNum;
+          break;
+        case 'D':
+          getScore = -1 * Math.pow(gottonNum, 2);
+          break;
+        case 'T':
+          getScore = -1 * Math.pow(gottonNum, 3);
+          break;
+      }
+      score.push(getScore);
+    }
+  });
+
+  return score.reduce((acc, num) => acc + num, 0);
+}
+

```

Comment:
제 리뷰를 토대로 리팩토링 해봤습니다. 참고만 해주세요 ㅎㅎ

```suggestion
function solution(dartResult) {
  const pattern = /[0-9]+[STD][*#]?/g;
  const array = dartResult.match(pattern);

  const BONUS = { S: 1, D: 2, T: 3 };
  const BONUSES = Object.keys(BONUS);
  const OPTION = { "*": 2, "#": -1 };
  const OPTIONS = Object.keys(OPTION);

  const score = [];

  array.forEach((el) => {
    const individual = el.split("");
    const divider = individual.at(-1);
    let getScore = 0;

    if (BONUSES.includes(divider)) {
      const gottonNum = parseInt(individual.slice(0, -1).join(""));
      getScore = Math.pow(gottonNum, BONUS[divider]);
    }

    if (OPTIONS.includes(divider)) {
      if (divider === "*") score[score.length - 1] *= 2;
      const bonus = individual.at(-2);
      const gottonNum = parseInt(individual.slice(0, -2).join(""));
      getScore = OPTION[divider] * Math.pow(gottonNum, BONUS[bonus]);
    }

    score.push(getScore);
  });

  return score.reduce((acc, num) => acc + num, 0);
}
```

Reply:
😧 fantastic..ㅋㅋ 제가 원했던게 이건데.. 누가 저렇게 짰죠..? ㅋㅋ
쿼카에서는 at(-1)이 안먹혀서 인덱스를 length -1 로 했었는데, 프로그래머스에서는 잘 되네요!
감사합니다 :-)